base_layer/core/src/block.rs

```diff
@@ -29,18 +29,66 @@ use crate::{
 };
 
 /// A Tari block. Blocks are linked together into a blockchain.
-pub struct Block {
+pub struct Block<'a> {
     pub header: BlockHeader,
-    pub body: AggregateBody,
+    pub body: AggregateBody<'a>,
 }
 
 /// The components of the block or transaction. The same struct can be used for either, since in Mimblewimble,
 /// cut-through means that blocks and transactions have the same structure.
-pub struct AggregateBody {
+pub struct AggregateBody<'a> {
     /// List of inputs spent by the transaction.
-    pub inputs: Vec<TransactionInput>,
+    pub inputs: Vec<TransactionInput<'a>>,
     /// List of outputs the transaction produces.
-    pub outputs: Vec<TransactionOutput>,
+    pub outputs: Vec<TransactionOutput<'a>>,
     /// Kernels contain the excesses and their signatures for transaction
-    pub kernels: Vec<TransactionKernel>,
+    pub kernels: Vec<TransactionKernel<'a>>,
+}
+
+impl<'a> AggregateBody<'a> {
+    /// Create an empty aggregate body
+    pub fn empty() -> AggregateBody<'a> {
+        AggregateBody { inputs: vec![], outputs: vec![], kernels: vec![] }
+    }
+
+    /// Create a new aggregate body from provided inputs, outputs and kernels
+    pub fn new(
+        inputs: Vec<TransactionInput<'a>>,
+        outputs: Vec<TransactionOutput<'a>>,
+        kernels: Vec<TransactionKernel<'a>>,
+    ) -> AggregateBody<'a>
+    {
+        AggregateBody { inputs, outputs, kernels }
+    }
+
+    /// Add an input to the existing aggregate body
+    pub fn with_input(self, input: TransactionInput<'a>) -> AggregateBody<'a> {
```

Comment:
Alright makes sense, this does couple these building functions to the final transaction. The next phase is to make a TransactionBuilder that manages the transaction negotiation process so these methods should probably end up in there.

Reply:
So I left the AggregateBody as is because it is only ever a component of a block or transaction and I moved the builder stuff into a TransactionBuilder and later we can make a BlockBuilder